packages/autocomplete-core/src/utils/createCancelablePromise.ts

```diff
@@ -0,0 +1,161 @@
+import { noop } from '@algolia/autocomplete-shared';
+
+type CancelablePromiseState = {
+  isCanceled: boolean;
+  onCancelList: Array<(...args: any[]) => any>;
+};
+
+type PromiseExecutor<TValue> = (
+  resolve: (value: TValue | PromiseLike<TValue>) => void,
+  reject: (reason?: any) => void
+) => void;
+
+type CreateInternalCancelablePromiseParams<TValue> = {
+  executor?: PromiseExecutor<TValue>;
+  promise?: Promise<TValue>;
+  initialState?: CancelablePromiseState;
+};
+
+function createInternalCancelablePromise<TValue>({
+  executor = noop,
+  initialState = createInitialState(),
+  promise = new Promise<TValue>((resolve, reject) => {
+    return executor(resolve, reject);
+  }),
+}: CreateInternalCancelablePromiseParams<TValue>): CancelablePromise<TValue> {
+  const state = initialState;
+
+  return {
+    then(onfulfilled, onrejected) {
+      return createCancelable(
+        promise.then(
+          createCallback(onfulfilled, state, promise),
+          createCallback(onrejected, state, promise)
+        ),
+        state
+      );
+    },
+    catch(onrejected) {
+      return createCancelable(
+        promise.catch(createCallback(onrejected, state, promise)),
+        state
+      );
+    },
+    finally(onfinally, runWhenCanceled) {
+      if (runWhenCanceled && onfinally) {
+        state.onCancelList.push(onfinally);
+      }
+
+      return createCancelable<TValue>(
+        promise.finally(
+          createCallback(
+            onfinally &&
+              (() => {
+                if (runWhenCanceled) {
+                  state.onCancelList = [];
+                }
+
+                return onfinally();
+              }),
+            state,
+            promise
+          )
+        ),
+        state
+      );
+    },
+    cancel() {
+      state.isCanceled = true;
+
+      state.onCancelList.forEach((callback) => {
+        callback();
+      });
+    },
+    isCanceled() {
+      return state.isCanceled === true;
+    },
+  };
+}
+
+export type CancelablePromise<TValue> = {
```

Comment:
This type is a copy of the `Promise` interface (slightly adapted for ESLint compliance and to the cancelable promise use case).